torch/__init__.py

```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
     Note that deterministic operations tend to have worse performance than
     non-deterministic operations.
 
+    .. note::
+
+        :func:`use_deterministic_algorithms` is not guaranteed to prevent
```

Comment:
If the indexing operations part of this note is removed, should this just warn about internal memory overlaps? Like:

"Inplace operations on tensors that index the same element multiple times may not be detected or prevented by this flag."

Follow-up question, do we consider different orders of floating point or complex arithmetic operations to be nondeterministic?

Reply:
Right, I'll only mention overlaps in this note.

Regarding order of operations, if the implementation of a given function could give a different order each time it's called in a consistent environment, then we should add an alert for it or offer an alternative implementation. We're already doing this--we considering `gpuAtomicAdd` nondeterministic for this exact reason.

For functions that do have a consistent internal order of operations, the order of operations is controlled by the user's code. I don't think we should alert in this case (I'm not even sure if/how we could detect something like this). In the following example, we're specifically asking for the sum to be performed with two different orders, and we get two different results. In short, if an order of operations difference is due to an actual user code change, it should still be considered deterministic.
```
>>> a = torch.randn(10000)
>>> a.sum().item()
6.353595733642578
>>> torch.stack([a[5000:], a[:5000]]).sum().item()
6.353599548339844
```

There's also the case of vectorized vs. nonvectorized operations, which effectively have different internal orders of operations. However, this is an environmental difference, so it should be considered deterministic as well.

Perhaps the reproducibility doc could mention some of these points (it already specifically mentions environmental differences, but not code differences), but I don't think the `use_deterministic_algorithms` doc is the right place for it. That is, unless there are some cases I'm not thinking of.

Reply:
I appreciate your thinking. Not elaborating here sounds good. Maybe we can think how to incorporate some ideas in the reproducibility note in the future.